### PR TITLE
Hadoop 16900 Very large files can be truncated when written through S3AFileSystem

### PR DESCRIPTION
**Pull request by mukund-thakur**
On top of https://github.com/apache/hadoop/pull/1944

Tested with ap-south-bucket. All good.

### CODE REVIEW

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3ABlocksize.java [hunk 1]
```diff
@@ -18,9 +18,12 @@
 
 package org.apache.hadoop.fs.s3a;
 
+import java.io.IOException;
+
 import org.apache.hadoop.fs.FileStatus;
 import org.apache.hadoop.fs.FileSystem;
 import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.s3a.impl.InternalConstants;
```
Comment (mukund-thakur) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3ABlocksize.java:26]: I think I will have to remove these import. Looks like got imported during experiments.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 27s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +0 :ok: |  markdownlint  |   0m  0s |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 3 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  21m 51s |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 31s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   0m 24s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 35s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  16m 11s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 26s |  trunk passed  |
| +0 :ok: |  spotbugs  |   1m  2s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   1m  0s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 34s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 36s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 36s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 18s |  hadoop-tools/hadoop-aws: The patch generated 7 new + 5 unchanged - 0 fixed = 12 total (was 5)  |
| +1 :green_heart: |  mvnsite  |   0m 34s |  the patch passed  |
| -1 :x: |  whitespace  |   0m  0s |  The patch has 1 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply  |
| -1 :x: |  whitespace  |   0m  0s |  The patch 5 line(s) with tabs.  |
| +1 :green_heart: |  shadedclient  |  15m 36s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 22s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   1m  3s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 25s |  hadoop-aws in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 27s |  The patch does not generate ASF License warnings.  |
|  |   |  63m 25s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1987/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1987 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle markdownlint |
| uname | Linux 342ed59b1f4f 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / db6252b6c39 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1987/1/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-1987/1/artifact/out/whitespace-eol.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-1987/1/artifact/out/whitespace-tabs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1987/1/testReport/ |
| Max. process+thread count | 342 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1987/1/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by steveloughran** [comment]
You shouldn't be using the same file to 2 uploads in a row, as it stops your making assertion that the destination file doesn't exist after the failed upload.

Because that is exactly the thing we need to make sure does not happen, alongside the actual exception being raised.

cherry picked the changes into my branch and extended the tests

**Comment by steveloughran** [comment]
cherry picked changes into my branch

**Pull request closed without merging**